src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1292 @@
+        The OU string read from top to bottom without RDNs and delimited by a "/". E.g. "createcomputer=Computers/Servers/Unix NB: A backslash
+        "\" is used as escape at multiple levels and may need to be doubled or even quadrupled. It is not used as a separator.
+        :idmap_backend: - The idmap backend provides a plugin interface for Winbind to use varying backends to store SID/uid/gid mapping tables.
+
+        The Active Directory service will be started after a configuration update if the service was initially disabled, and the updated
+        configuration sets :enable: to True. Likewise, the Active Directory service will be stopped if :enable: is changed to False
+        during an update. If the configuration is updated, but the initial :enable: state was True, then only a samba_server restart
+        command will be issued.
+        """
+        old = await self.config()
+        new = old.copy()
+        new.update(data)
+        new = await self.ad_compress(new)
+        await self.middleware.call(
+            'datastore.update',
+            'directoryservice.activedirectory',
+            old['id'],
+            new,
+            {'prefix': 'ad_'}
+        )
+        verrors = ValidationErrors()
+        if not new["bindname"] and not new["kerberos_principal"]:
+            verrors.add(f"activedirectory_update.bindname", "Bind credentials or kerberos keytab are required to join an AD domain.")
+
+        if data['bindpw'] and data['enable'] and not old['enable']:
+            try:
+                await self.middleware.run_in_thread(self.validate_credentials)
+            except Exception as e:
+                verrors.add(f"activedirectory_update.bindpw", f"Failed to validate bind credentials: {e}")
+
+            await self.middleware.run_in_thread(self.validate_domain)
+
+        if verrors:
+            raise verrors
+
+        start = False
+        stop = False
+
+        if old['idmap_backend'] != new['idmap_backend']:
+            idmap = await self.middleware.call('idmap.domaintobackend.query', [('domain', '=', 'DS_TYPE_ACTIVEDIRECTORY')])
+            self.logger.debug(idmap[0]['id'])
+            await self.middleware.call('idmap.domaintobackend.update', idmap[0]['id'], {'idmap_backend': new['idmap_backend']})
+
+        if not old['enable']:
+            if new['enable']:
+                start = True
+        else:
+            if not new['enable']:
+                stop = True
+
+        if stop:
+            await self.stop()
+        if start:
+            await self.start()
+
+        if not stop and not start and new['enable']:
+            await self.middleware.call('service.restart', 'cifs')
+
+        return await self.config()
+
+    @private
+    async def _set_state(self, state):
+        await self.middleware.call('cache.put', 'AD_State', state.name)
+
+    @private
+    async def get_state(self):
```

Comment:
Its weird that _set_state starts with underscode but get_state doesnt? : P

Reply:
get_state() shouldn't be private. So I'll fix that.